build.sbt

```diff
@@ -200,14 +182,50 @@ lazy val commonSettings = Seq(
   testOptions += Tests.Argument(TestFrameworks.JUnit, "-q", "--verbosity=1"),
 ) ++ Defaults.itSettings
 
-def scalacCrossOptions(scalaVersion: String) =
-  CrossVersion.partialVersion(scalaVersion) match {
+
+def buildScalacOptions(scalaVersion: String) = {
+  val commonOptions = Seq(
+    "-target:jvm-1.8",
+    "-feature",
+    "-deprecation",
+    "-language:experimental.macros",
+    "-unchecked",
+    "-Xfatal-warnings",
+    "-Xxml:-coalescing",
+    "-Xfuture"
+  )
+
+  val scalaVersionSpecificOptions = CrossVersion.partialVersion(scalaVersion) match {
     case Some((2, 12)) => Seq(
       "-Ywarn-unused:imports"
     )
     case _ => Seq.empty
   }
 
+  val javaVersionSpecificOptions =
+    if (scala.util.Properties.isJavaAtLeast("9"))
+      Seq("-release", "8")
+    else
+      Seq.empty
+
+    commonOptions ++ scalaVersionSpecificOptions ++ javaVersionSpecificOptions
+}
+
+def buildJavacOptions() = {
+  val commonOptions = Seq(
+    "-Werror",
+    "-Xlint:deprecation"
+  )
+
+  val javaVersionSpecificOptions =
```

Comment:
Can you please just add comments to this code change - can just include a hint about the issue + a link to the JIRA issue discussion if you want. I just want this complexity in here to be well motivated.